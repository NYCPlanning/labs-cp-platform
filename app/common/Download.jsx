// Download.jsx - This component builds a download pane used in the explorer

import React, { PropTypes } from 'react';

import Subheader from 'material-ui/Subheader';
import FlatButton from 'material-ui/FlatButton';
import FontIcon from 'material-ui/FontIcon';

import Carto from '../helpers/carto';

import Moment from 'moment';

const date = Moment().format('YYYY-MM-DD');

const Download = ({ sql, filePrefix = 'data' }) => {
  const filteredSql = Carto.transformSqlSelectAll(sql);
  const completeSql = Carto.transformSqlRemoveWhere(filteredSql);

  const allFilename = `${filePrefix}_all_${date}`;
  const filteredFilename = `${filePrefix}_filtered_${date}`;

  return (
    <div>
      <Subheader>
        Complete Dataset
      </Subheader>

      <FlatButton
        label="CSV"
        href={Carto.generateUrlString(completeSql, 'csv', allFilename)}
        download="text.csv"
        icon={<FontIcon className="fa fa-file-excel-o" style={{ fontSize: '14px' }} />}
      />
      <FlatButton
        label="GeoJSON"
        href={Carto.generateUrlString(completeSql, 'geojson', allFilename)}
        icon={<FontIcon className="fa fa-file-code-o" style={{ fontSize: '14px' }} />}
      />
      <FlatButton
        label="Shapefile"
        href={Carto.generateUrlString(completeSql, 'shp', allFilename)}
        icon={<FontIcon className="fa fa-file-archive-o" style={{ fontSize: '14px' }} />}
      />

      <Subheader>
        Filtered Dataset
      </Subheader>

      <FlatButton
        label="CSV"
        href={Carto.generateUrlString(filteredSql, 'csv', filteredFilename)}
        icon={<FontIcon className="fa fa-file-excel-o" style={{ fontSize: '14px' }} />}
      />
      <FlatButton
        label="GeoJSON"
        href={Carto.generateUrlString(filteredSql, 'geojson', filteredFilename)}
        icon={<FontIcon className="fa fa-file-code-o" style={{ fontSize: '14px' }} />}
      />
      <FlatButton
        label="Shapefile"
        href={Carto.generateUrlString(filteredSql, 'shp', filteredFilename)}
        icon={<FontIcon className="fa fa-file-archive-o" style={{ fontSize: '14px' }} />}
      />
    </div>
  );
};

Download.propTypes = {
  sql: PropTypes.string,
  filePrefix: PropTypes.string,
};

module.exports = Download;
