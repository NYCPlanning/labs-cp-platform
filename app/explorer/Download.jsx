import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import DownloadButton from './download/DownloadButton';

class Download extends React.Component {
  activeLayers = () => this.props.layers.filter(layer => layer.enabled);
  activeLayerIDs = () => this.activeLayers().map(layer => layer.id);

  layerMap = {
    'capital-projects': {
      title: 'Capital Projects',
      noun: 'projects',
      icon: 'usd',
    },
    scaplan: {
      title: 'SCA Capital Projects',
      noun: 'projects',
      icon: 'graduation-cap',
    },
    'housing-development': {
      title: 'Housing Development',
      noun: 'developments',
      icon: 'cubes',
    },
    'cb-budgetrequests': {
      title: 'Budget Requests',
      noun: 'requests',
      icon: 'book',
    },
    'facilities-cp': {
      title: 'Facilities',
      noun: 'facilities',
      icon: 'university',
    },
  }

  layers = [
    'capital-projects',
    // 'scaplan',
    'housing-development',
    'cb-budgetrequests',
    'facilities-cp',
  ];

  render() {
    const { counts, sql } = this.props;
    const layers = this.props.isLoggedIn ? this.layers : ['facilities-cp'];

    return (<div>
      {
        layers.map(layerID => (
          <div className="download-row" key={layerID}>
            <h5><span className={`fa fa-${this.layerMap[layerID].icon}`} /> {this.layerMap[layerID].title}</h5>
            <DownloadButton
              layerID={layerID}
              noun={this.layerMap[layerID].noun}
              sql={sql[layerID]}
              counts={{
                total: counts.total[layerID],
                filtered: counts.filtered[layerID],
              }}
              filtered={counts.total[layerID] !== counts.filtered[layerID]}
            />
          </div>
        ))
      }
    </div>);
  }
}

Download.propTypes = {
  layers: PropTypes.array.isRequired,
  counts: PropTypes.object.isRequired,
  sql: PropTypes.object.isRequired,
  isLoggedIn: PropTypes.bool.isRequired,
};


const mapStateToProps = ({ facilitiesCP, capitalProjects, cbBudgetRequests, housingDevelopment, currentUser }) => ({
  counts: {
    total: {
      'capital-projects': capitalProjects.pointsTotalCount + capitalProjects.polygonsTotalCount,
      'facilities-cp': facilitiesCP.totalCount,
      'cb-budgetrequests': cbBudgetRequests.totalCount,
      'housing-development': housingDevelopment.totalCount,
    },
    filtered: {
      'capital-projects': capitalProjects.selectedCount,
      'facilities-cp': facilitiesCP.selectedCount,
      'cb-budgetrequests': cbBudgetRequests.selectedCount,
      'housing-development': housingDevelopment.selectedCount,
    },
  },
  sql: {
    'capital-projects': capitalProjects.sql,
    'facilities-cp': facilitiesCP.sql,
    'cb-budgetrequests': cbBudgetRequests.sql,
    'housing-development': housingDevelopment.sql,
  },
  isLoggedIn: currentUser.isLoggedIn,
});

export default connect(mapStateToProps)(Download);
