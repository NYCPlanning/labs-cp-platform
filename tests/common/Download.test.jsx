import renderer from 'react-test-renderer';
import React from 'react';
import Download from '../../app/common/Download';

// Boilerplate required by MUI at the moment :/
// https://github.com/callemall/material-ui/issues/686
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
const muiTheme = getMuiTheme();

test('Download.jsx does its thing', () => {
  Date.now = jest.fn(() => 1492014178000);

  const component = renderer.create(
    <MuiThemeProvider muiTheme={muiTheme}>
      <Download
        sql="SELECT * FROM that_table"
        filePrefix="good_stuff"
      />
    </MuiThemeProvider>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
