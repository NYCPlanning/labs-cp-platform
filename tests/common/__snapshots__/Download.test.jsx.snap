// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Download.jsx does its thing 1`] = `
<div>
  <h4>
    Complete Dataset
  </h4>
  <a
    disabled={false}
    download="text.csv"
    href="https://cartoprod.capitalplanning.nyc/user/cpp/api/v2/sql?q=SELECT%20*%20FROM%20that_table&format=csv&filename=good_stuff_complete_2017-04-12"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "border": 10,
        "borderRadius": 2,
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "cursor": "pointer",
        "display": "inline-block",
        "fontFamily": "Roboto, sans-serif",
        "fontSize": "inherit",
        "fontWeight": "inherit",
        "height": 36,
        "lineHeight": "36px",
        "margin": 0,
        "minWidth": 88,
        "muiPrepared": true,
        "outline": "none",
        "overflow": "hidden",
        "padding": 0,
        "position": "relative",
        "textAlign": "center",
        "textDecoration": "none",
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "userSelect": "none",
        "verticalAlign": "middle",
        "zIndex": 1,
      }
    }
    tabIndex={0}
  >
    <div
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="fa fa-file-excel-o"
        color="rgba(0, 0, 0, 0.87)"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "display": "inline-block",
            "fontSize": "11px",
            "marginLeft": 12,
            "marginRight": 0,
            "muiPrepared": true,
            "position": "relative",
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "userSelect": "none",
            "verticalAlign": "middle",
          }
        }
      />
      <span
        style={
          Object {
            "fontSize": "11px",
            "fontWeight": 500,
            "letterSpacing": 0,
            "muiPrepared": true,
            "paddingLeft": 8,
            "paddingRight": 16,
            "position": "relative",
            "textTransform": "uppercase",
            "verticalAlign": "middle",
          }
        }
      >
        CSV
      </span>
    </div>
  </a>
  <a
    disabled={false}
    href="https://cartoprod.capitalplanning.nyc/user/cpp/api/v2/sql?q=SELECT%20*%20FROM%20that_table&format=geojson&filename=good_stuff_complete_2017-04-12"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "border": 10,
        "borderRadius": 2,
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "cursor": "pointer",
        "display": "inline-block",
        "fontFamily": "Roboto, sans-serif",
        "fontSize": "inherit",
        "fontWeight": "inherit",
        "height": 36,
        "lineHeight": "36px",
        "margin": 0,
        "minWidth": 88,
        "muiPrepared": true,
        "outline": "none",
        "overflow": "hidden",
        "padding": 0,
        "position": "relative",
        "textAlign": "center",
        "textDecoration": "none",
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "userSelect": "none",
        "verticalAlign": "middle",
        "zIndex": 1,
      }
    }
    tabIndex={0}
  >
    <div
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="fa fa-file-code-o"
        color="rgba(0, 0, 0, 0.87)"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "display": "inline-block",
            "fontSize": "11px",
            "marginLeft": 12,
            "marginRight": 0,
            "muiPrepared": true,
            "position": "relative",
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "userSelect": "none",
            "verticalAlign": "middle",
          }
        }
      />
      <span
        style={
          Object {
            "fontSize": "11px",
            "fontWeight": 500,
            "letterSpacing": 0,
            "muiPrepared": true,
            "paddingLeft": 8,
            "paddingRight": 16,
            "position": "relative",
            "textTransform": "uppercase",
            "verticalAlign": "middle",
          }
        }
      >
        GeoJSON
      </span>
    </div>
  </a>
  <a
    disabled={false}
    href="https://cartoprod.capitalplanning.nyc/user/cpp/api/v2/sql?q=SELECT%20*%20FROM%20that_table&format=shp&filename=good_stuff_complete_2017-04-12"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "border": 10,
        "borderRadius": 2,
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "cursor": "pointer",
        "display": "inline-block",
        "fontFamily": "Roboto, sans-serif",
        "fontSize": "inherit",
        "fontWeight": "inherit",
        "height": 36,
        "lineHeight": "36px",
        "margin": 0,
        "minWidth": 88,
        "muiPrepared": true,
        "outline": "none",
        "overflow": "hidden",
        "padding": 0,
        "position": "relative",
        "textAlign": "center",
        "textDecoration": "none",
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "userSelect": "none",
        "verticalAlign": "middle",
        "zIndex": 1,
      }
    }
    tabIndex={0}
  >
    <div
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="fa fa-file-archive-o"
        color="rgba(0, 0, 0, 0.87)"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "display": "inline-block",
            "fontSize": "11px",
            "marginLeft": 12,
            "marginRight": 0,
            "muiPrepared": true,
            "position": "relative",
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "userSelect": "none",
            "verticalAlign": "middle",
          }
        }
      />
      <span
        style={
          Object {
            "fontSize": "11px",
            "fontWeight": 500,
            "letterSpacing": 0,
            "muiPrepared": true,
            "paddingLeft": 8,
            "paddingRight": 16,
            "position": "relative",
            "textTransform": "uppercase",
            "verticalAlign": "middle",
          }
        }
      >
        Shapefile
      </span>
    </div>
  </a>
  <h4>
    Filtered Dataset
  </h4>
  <a
    disabled={false}
    href="https://cartoprod.capitalplanning.nyc/user/cpp/api/v2/sql?q=SELECT%20*%20FROM%20that_table&format=csv&filename=good_stuff_filtered_2017-04-12"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "border": 10,
        "borderRadius": 2,
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "cursor": "pointer",
        "display": "inline-block",
        "fontFamily": "Roboto, sans-serif",
        "fontSize": "inherit",
        "fontWeight": "inherit",
        "height": 36,
        "lineHeight": "36px",
        "margin": 0,
        "minWidth": 88,
        "muiPrepared": true,
        "outline": "none",
        "overflow": "hidden",
        "padding": 0,
        "position": "relative",
        "textAlign": "center",
        "textDecoration": "none",
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "userSelect": "none",
        "verticalAlign": "middle",
        "zIndex": 1,
      }
    }
    tabIndex={0}
  >
    <div
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="fa fa-file-excel-o"
        color="rgba(0, 0, 0, 0.87)"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "display": "inline-block",
            "fontSize": "11px",
            "marginLeft": 12,
            "marginRight": 0,
            "muiPrepared": true,
            "position": "relative",
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "userSelect": "none",
            "verticalAlign": "middle",
          }
        }
      />
      <span
        style={
          Object {
            "fontSize": "11px",
            "fontWeight": 500,
            "letterSpacing": 0,
            "muiPrepared": true,
            "paddingLeft": 8,
            "paddingRight": 16,
            "position": "relative",
            "textTransform": "uppercase",
            "verticalAlign": "middle",
          }
        }
      >
        CSV
      </span>
    </div>
  </a>
  <a
    disabled={false}
    href="https://cartoprod.capitalplanning.nyc/user/cpp/api/v2/sql?q=SELECT%20*%20FROM%20that_table&format=geojson&filename=good_stuff_filtered_2017-04-12"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "border": 10,
        "borderRadius": 2,
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "cursor": "pointer",
        "display": "inline-block",
        "fontFamily": "Roboto, sans-serif",
        "fontSize": "inherit",
        "fontWeight": "inherit",
        "height": 36,
        "lineHeight": "36px",
        "margin": 0,
        "minWidth": 88,
        "muiPrepared": true,
        "outline": "none",
        "overflow": "hidden",
        "padding": 0,
        "position": "relative",
        "textAlign": "center",
        "textDecoration": "none",
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "userSelect": "none",
        "verticalAlign": "middle",
        "zIndex": 1,
      }
    }
    tabIndex={0}
  >
    <div
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="fa fa-file-code-o"
        color="rgba(0, 0, 0, 0.87)"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "display": "inline-block",
            "fontSize": "11px",
            "marginLeft": 12,
            "marginRight": 0,
            "muiPrepared": true,
            "position": "relative",
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "userSelect": "none",
            "verticalAlign": "middle",
          }
        }
      />
      <span
        style={
          Object {
            "fontSize": "11px",
            "fontWeight": 500,
            "letterSpacing": 0,
            "muiPrepared": true,
            "paddingLeft": 8,
            "paddingRight": 16,
            "position": "relative",
            "textTransform": "uppercase",
            "verticalAlign": "middle",
          }
        }
      >
        GeoJSON
      </span>
    </div>
  </a>
  <a
    disabled={false}
    href="https://cartoprod.capitalplanning.nyc/user/cpp/api/v2/sql?q=SELECT%20*%20FROM%20that_table&format=shp&filename=good_stuff_filtered_2017-04-12"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "border": 10,
        "borderRadius": 2,
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "cursor": "pointer",
        "display": "inline-block",
        "fontFamily": "Roboto, sans-serif",
        "fontSize": "inherit",
        "fontWeight": "inherit",
        "height": 36,
        "lineHeight": "36px",
        "margin": 0,
        "minWidth": 88,
        "muiPrepared": true,
        "outline": "none",
        "overflow": "hidden",
        "padding": 0,
        "position": "relative",
        "textAlign": "center",
        "textDecoration": "none",
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "userSelect": "none",
        "verticalAlign": "middle",
        "zIndex": 1,
      }
    }
    tabIndex={0}
  >
    <div
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="fa fa-file-archive-o"
        color="rgba(0, 0, 0, 0.87)"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "display": "inline-block",
            "fontSize": "11px",
            "marginLeft": 12,
            "marginRight": 0,
            "muiPrepared": true,
            "position": "relative",
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "userSelect": "none",
            "verticalAlign": "middle",
          }
        }
      />
      <span
        style={
          Object {
            "fontSize": "11px",
            "fontWeight": 500,
            "letterSpacing": 0,
            "muiPrepared": true,
            "paddingLeft": 8,
            "paddingRight": 16,
            "position": "relative",
            "textTransform": "uppercase",
            "verticalAlign": "middle",
          }
        }
      >
        Shapefile
      </span>
    </div>
  </a>
</div>
`;
